Adding maptracker logic

Levisaxos opened the pull request:
Added full maptracker support based on apworld logic.
Keystones are not considered in logic for certain locations untill you have a lot of them.
So it might be possible to get to items using keystones while the tracker may say that are not reachable. If so, these items are not considered in logic by archipelago nor the tracker.
There should also be icons for checks that are normally not visible (Horu doors, escape scenes etc).
In settings you are set your icons and mapvisibility to what you want.

Work to do / known issues: 
- Misty Woods
- End escape in horu
- ~~There might be some weird things going on when switching between slots.  Unless you want to test this and see whats wrong I advice to always close the game before starting another slot.~~ Fixed this
- Ability pedestals and save points icons remain active even after capturing them making it difficult to see which you havn't checked yet.
- Logic is based off ingame check collected. Meaning, if you die, archiepelago will still give you the item but the maptracker will say the location is not collected untill you save.

Some QOL fixes:
- Showing archipelago input fields only on save form
- Added nullchecks for server, port and slotname.
- Added filling of server, port and slotname on saveui entry so first slot can be started without having to go to second slot and then back to first slot.
- Added a teleport to start to avoid softlocks when forgetting to take sunken glades TP

c-ostic commented:
First of all, great work! I know the logic for Ori is a complex mess, but kudos to you on implementing the parser anyway. And I appreciate the smaller fixes you made along the way.

I took a look at your code, and also was able to briefly test it in-game. I have a few comments, questions, and issues to address before merging everything in.

Comments
1. I really like your ModLogger class. I've been meaning to add a more official logger into the mod, but hadn't gotten around to it. One thing I would add is for it to log to the BepinEx console in addition to writing to the file. Also, I noticed a comment in the Initialize function about creating a file with timestamp, though that doesn't seem to be the case right now.
2. Great job with the tracker settings as well. I may emulate your class for the randomizer settings I have if you don't mind.
3. I also like the organizational structure you have in place for all of the map tracker files and would be willing to hear any ideas about organizing some of the other files in the project (if you have any). Regarding ModLogger.cs and BaseModOptionsScreen.cs in particular, would it make sense to move these into a different folder since they are not unique to the map tracker?
4. Thank you for fixing some of the smaller issues with the archipelago slot data ui, as well as adding the teleport to start function

Questions
1. As far as I can tell, LogicInventory seems to be a proxy between LogicManager and RandomizerReceiver. What is the purpose of this layer of separation?
2. I noticed that the UI for RandomizerSettings has been changed to only appear on the map instead of also on the pause menu. Was there a particular reason for this?

Issues
1. I noticed in the LogicInventory it looks like sorrowPass has been given TPValley instead of TPSorrow, while valleyOfTheWind was given nothing.
2. In the short amount of time that I tested the game, I noticed that some of the checks that should be in logic weren't until I picked up Sein. Likely not a huge issue (since this is the first thing players do anyway), but something to keep in mind
3. The map tracker ui that shows up when hovering over a location does not seem to scale with screen size (or is otherwise much smaller than the other text)

I know this is a lot all at once, but feel free to address these at your leisure. Once again, awesome work on this! And of course, feel free to ask me any questions yourself.

Levisaxos commented:
Incoming wall of text!

Great. Thank you for reviewing.
Tbh I tried a lot of things with the logic, even implementing it by hand based on the poptracker but I couldn't manage to get it working properly so eventually I popped all logic and the python Rules.py into AI and asked it to create the checker for me.
With a lot of follow up questions and some hand changes it finally worked as intended.

Thank you for the compliments on the ModLogger. I believe logging should be done in seperate places and not duplicate, so in my opinion logging to bepinex AND mod log files shouldn't be done.
Also the "enable debug" option can not be set ingame, only by editing the config file on your computer so the debug logs are not shown in users clients. Also good call on the timestamp. I did originally implement that but at some point it might have gone lost in all the copy pasting, so i'll add that again.

For now I've kept a clear diffence between the archipelago classes and the maptracker classes (except for the patches). This is mostly due to not having asked yet what changes I can and can not make in the archipelago classes.
So I think a bit of refacting on the structure, combining some classes that do the same thing and moving some files around should be in order. 

Now about your questions and issues:
Q1:  When i started merging my project into this I didn't look at what was already there so I just copied my LogicInventory into the archpelago project and just collected the other settings I needed from the archipelago classes.
Now that I look at it a bit more, I feel my LogicInventory is basicly the same as your ArchipelagoInventory, so they could be combined.
There's no reason why I feel they should be seperate other than this was the easiest way for me to implement it without having to rewrite both classes.

Q2:  This was a personal preference, these settings are only meant to be changed in game and for me it feels more logical to have them only visible on the map menu. 
Ofcourse, in the future I feel these should also move to an actual settingssscreen instead of on the mapmenu.

i1: I actually didn't notice that yet in my games. I'll have to look into that and see what i'm missing in these mappings. Thank you for noticing.

i2: This is actually intentional. I added another layer in the maptracker logic since the items that are shown as in logic (GladesKeystones), aren't actually in logic untill you get sein due to the damaging eggplants (or whatever they are called).
However I have managed to get an damaging item (like chargejump) on the first item which in turn would put them in logic, so I might have been wrong in this case.
My reasoning is the same as that I once hardlocked myself by not taking the sunken glades teleporter. It is to avoid missing something important (sein) before moving on to other check that require sein (or another damaging skill).

i3: This might be true that I set it to a lower fontsize that usual. It was one of the first things I made and didn't change a lot to the styling of it.

After creating the pull request I continued working on a few things, like I noticed that some of the custom icon (Icons that are not normally in the game, like the events and plants) would not properly hide on collection.
There is still the issue for plants that I can only mark them as collected during collection.  I've not found a way yet to determine wether a plant is collected or not from a loaded game, so i'm using a file to log which customicon is collected by moonguid.

I've also added the goal icon marker to the game. Still have to test if it will show as in logic when i have all requirements but havn't gotten that far in any of my games yet.
These, along with some other small fixes and QOL improvements, are currently in another branch which I could merge back into this one
.
Not sure what the plan is for now, my idea was mainly to get a testing version out for the public to use, not specifically have a release version yet.
My experience with PR's is asfar that at my work any and all PR's are quickly glanced at and immediatly accepted (barring stupid mistakes).
So perhaps you could explain it a bit more to me? Is the intention to have it fully functioning before accepting a PR and is a PR accepted considered a new version that people can download? 
If it's all the same to you, we can discuss some of the points/issues in discord via private messages, otherwise the posts here can get massive.

Levisaxos commented:
I've checked the teleporters and it is true that sorrowpass and valleyofthewind teleporters are mixed up:
[13:45:05 Debug] Teleporting to valleyOfTheWind goes to Sorrow Pass
[13:45:33 Debug] Teleporting to sorrowPass goes to Valley Of The Wind

[13:44:51 Debug] Teleporting to forlorn goes to Forlorn Ruins
[13:45:18 Debug] Teleporting to mountHoru goes to Mount Horu
[13:45:45 Debug] Teleporting to spiritTree goes to Hollow Grove
[13:45:58 Debug] Teleporting to sunkenGlades goes to Sunken Glades
[13:46:12 Debug] Teleporting to ginsoTree goes to Ginso Tree
[13:46:23 Debug] Teleporting to swamp goes to Thornfelt Swamp
[13:46:37 Debug] Teleporting to moonGrotto goes to Moon Grotto
[13:46:50 Debug] Teleporting to mangroveFalls goes to Mangrove area


I have yet to unlock the horuFields teleporter, but will look into that.

c-ostic commented:
After about 2 weeks of having a playtest version with no major game-breaking bugs, and some other logic issues fixed, I think this is good to be merged into the main project. Great work!
